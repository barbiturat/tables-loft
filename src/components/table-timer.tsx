import * as React from 'react';
import { connect } from 'react-redux';
import * as moment from 'moment';
import { compose, withHandlers, withState } from 'recompose';
import * as R from 'ramda';

import { StoreStructure } from '../interfaces/store-models';

interface Props {
  readonly isActive: boolean;
  readonly startsAt?: number;
}

interface MappedProps {
  readonly utcMilliseconds: number;
}

const getDurationActivityString = (
  isFormatOfMinutes: boolean,
  utcMilliseconds: number,
  startsAt?: number
): string => {
  if (!startsAt || !utcMilliseconds) {
    return 'Wrong Parameters!';
  }

  const utcMillisecondsFixed = Math.max(startsAt, utcMilliseconds);
  const durationMs = utcMillisecondsFixed - startsAt;

  if (isFormatOfMinutes) {
    const duration = moment.duration(durationMs);
    const minutes = Math.floor(duration.asMinutes());

    return `${minutes}m`;
  } else {
    return moment.utc(durationMs).format('H[h] mm[m] ss[s]');
  }
};

const enhance = compose(
  withState('isFormatOfMinutes', 'setFormatOfMinutes', false),
  withHandlers({
    onClick: ({ isFormatOfMinutes, setFormatOfMinutes }) => () =>
      setFormatOfMinutes(isFormatOfMinutes)
  })
);

const Component = enhance(
  ({ isActive, isFormatOfMinutes, utcMilliseconds, startsAt, onClick }: any) => {
    const getIsAvailable = () => !isActive;
    const getLabelAvailableText = R.ifElse(
      getIsAvailable,
      R.always('Available'),
      R.partial(getDurationActivityString, [isFormatOfMinutes, utcMilliseconds, startsAt])
    );
    const getAvailabilityClass = R.ifElse(
      getIsAvailable,
      R.always('table__label table__label_role_availability'),
      R.always('table__label table__label_role_counter')
    );

    return (
      <div className={getAvailabilityClass(null)} onClick={onClick}>
        {getLabelAvailableText(null)}
      </div>
    );
  }
);

const TableTimer = connect<
  any,
  any,
  Props
>((state: StoreStructure, ownProps: Props): MappedProps => {
  return {
    utcMilliseconds: state.app.utcMilliseconds
  };
})(Component);

export default TableTimer;
