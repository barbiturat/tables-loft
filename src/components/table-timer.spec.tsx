import * as React from 'react';
import * as renderer from 'react-test-renderer';

import {Component as TableTimerComponent} from './table-timer';

test('TableTimer renders correctly', () => {
  const now = 1491581751312;
  const isActive = true;

  const tree = (renderer as any).create(
    <TableTimerComponent
      isActive={isActive}
      startsAt={now}
    />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
