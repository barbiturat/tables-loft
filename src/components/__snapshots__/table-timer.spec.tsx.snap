// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<div
  className="table__label table__label_role_counter"
  onClick={[Function]}
>
  0h 00m 00s
</div>
`;

exports[`Renders correctly 2`] = `
<div
  className="table__label table__label_role_counter"
  onClick={[Function]}
>
  0m
</div>
`;
